### PR TITLE
Add proper TypeScript build logic

### PR DESCRIPTION
**Pull request by willarmiros**
*Issue #, if available:*
#411

*Description of changes:*
These changes allow the repo to conform to TypeScript best practices of not committing transpiled code to source control as we were doing before. Now, we have a `compile` target on the core package which should be run before testing or publishing the package. It transpiles the TypeScript in the source package, and copies the transpiled JavaScript and all other non-TS source files to a `dist` directory, which will ultimately be tested or vended to customers.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/pr-build.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ on:
   push:
     branches: [master]
   pull_request:
-    branches: [master]
+    branches: [master, aws-v3-support]
```
Comment (anuraaga) [inline .github/workflows/pr-build.yml:7]: @willarmiros Can you switch this PR to draft?

Comment (willarmiros) [reply .github/workflows/pr-build.yml:7]: Ah good call, sorry for the spam

**Comment by willarmiros** [comment]
@anuraaga this should now *actually* be ready for review if you wanna take a look :)

File: .github/workflows/pr-build.yml (hunk 1 above)
Comment (anuraaga) [inline .github/workflows/pr-build.yml:7]:
```suggestion
    branches: [master]
```

**Review by anuraaga — COMMENTED** [review]
> not committing transpiled code to source control 

Will you remove the current `.d.ts` files in another PR?

**Comment by willarmiros** [comment]
> Will you remove the current .d.ts files in another PR?

@anuraaga I wasn't sure whether or not I should. We have manually created `.d.ts` files and committed them to source control in all of our packages (not just core). I know the existing declaration files are correct, and was suspect of the generated ones. For instance, in the generated `aws-xray.d.ts`, where we're supposed to declare all of the top-level APIs of the SDK, this was all it contained:

```ts
export namespace middleware {
    const IncomingRequestData: typeof import("./middleware/incoming_request_data");
}
```

I'm sure we can make a series of changes to our source to make the generated declaration files work, but given that TS customers have been happy with the existing ones I'm not sure I'll prioritize it along with these changes.

**Comment by willarmiros** [comment]
I'm going to keep CI enabled on the `aws-v3-support` and go ahead and merge this so I can incorporate these changes with #416. I'll disable CI and we can re-review when I make the PR for merging `aws-v3-support` into master.

**Comment by anuraaga** [comment]
Oh ok if those are manual b declarations than sounds good didn't quite understand what's going on.

**Pull request merged**